packages/connection/legacy/class-jetpack-signature.php

```diff
@@ -325,20 +325,40 @@ public function encode_3986( $string_or_array ) {
 
 	/**
 	 * Concatenates a parameter name and a parameter value with an equals sign between them.
-	 * Supports one-dimensional arrays as `$value`.
 	 *
-	 * @param string $name  Parameter name.
-	 * @param mixed  $value Parameter value.
-	 * @return string A pair with parameter name and value (e.g. `name=value`).
+	 * @param string       $name  Parameter name.
+	 * @param string|array $value Parameter value.
+	 * @return string|array A string pair (e.g. `name=value`) or an array of string pairs.
 	 */
 	public function join_with_equal_sign( $name, $value ) {
 		if ( is_array( $value ) ) {
-			$result = array();
-			foreach ( $value as $array_key => $array_value ) {
-				$result[] = $name . '[' . $array_key . ']=' . $array_value;
-			}
-			return $result;
+			return $this->join_array_with_equal_sign( $name, $value );
 		}
 		return "{$name}={$value}";
 	}
+
+	/**
+	 * Helper function for join_with_equal_sign for handling arrayed values.
+	 * Explicitly supports nested arrays.
+	 *
+	 * @param string $name  Parameter name.
+	 * @param array  $value Parameter value.
+	 * @return array An array of string pairs (e.g. `[ name[example]=value ]`).
+	 */
+	private function join_array_with_equal_sign( $name, $value ) {
+		$result = array();
+		foreach ( $value as $value_key => $value_value ) {
+			$joined_value = $this->join_with_equal_sign( $name . '[' . $value_key . ']', $value_value );
+			if ( is_array( $joined_value ) ) {
+				foreach ( array_values( $joined_value ) as $individual_joined_value ) {
+					$result[] = $individual_joined_value;
+				}
+			} elseif ( is_string( $joined_value ) ) {
+				$result[] = $joined_value;
+			}
+		}
+
+		sort( $result );
```

Comment:
I think it'd make sense to remove this `sort()`, remove the `sort()` in `->normalized_query_parameters()` and add a `sort()` after the `$normalized_request_pieces = $flat_normalized_request_pieces` line in `->sign_request()`.

Otherwise, we're sorting string and array values together, which may work but I don't trust to keep working.

Otherwise, makes sense to me.

Reply:
It looks like performing a regular sort on `$normalized_request_pieces` is nondeterministic depending on PHP versions -- I sav discrepancies between `7.3.21` and `7.4.6`.

I was able to get around this by forcing [a string sort](https://www.php.net/manual/en/function.sort.php) on `$normalized_request_pieces`.

Reply:
Nice find. String sort is definitely what's intended here. What discrepancies were you finding?

Reply:
Sort of related - can we handle arbitrary array nesting depth with this change? I think the flattening step in `->sign_request()` needs some work too before we can.

Reply:
Here's the output of `sort` on 7.4.6:
```php
[31-Aug-2020 21:45:38 UTC] [de8d-270] Array
(
    [0] =>
    [1] => /rest/v1.3/sites/160291064/search
    [2] => GET
    [3] => K2BxdEKvZo
    [4] => fields[0]=date
    [5] => fields[1]=permalink.url.raw
    [6] => fields[2]=tag.name.default
    [7] => fields[3]=category.name.default
    [8] => fields[4]=post_type
    [9] => fields[5]=has.image
    [10] => fields[6]=shortcode_types
    [11] => fields[7]=image.url.raw
    [12] => highlight_fields[0]=title
    [13] => highlight_fields[1]=content
    [14] => highlight_fields[2]=comments
    [15] => public-api.wordpress.com
    [16] => query=hello
    [17] => size=10
    [18] => sort=score_default
    [19] => w^BXg^rmuzai3iBGaV$*wANJ1n*Y3uzq:1:0
    [20] => 443
    [21] => 1598910338
)
```

Whereas here's the output of `sort` from 7.3.21:
```php
[31-Aug-2020 21:45:38 UTC] [4bab-24233] Array
(
    [0] =>
    [1] => /rest/v1.3/sites/160291064/search
    [2] => 1598910338
    [3] => GET
    [4] => K2BxdEKvZo
    [5] => fields[0]=date
    [6] => fields[1]=permalink.url.raw
    [7] => fields[2]=tag.name.default
    [8] => fields[3]=category.name.default
    [9] => fields[4]=post_type
    [10] => fields[5]=has.image
    [11] => fields[6]=shortcode_types
    [12] => fields[7]=image.url.raw
    [13] => highlight_fields[0]=title
    [14] => highlight_fields[1]=content
    [15] => highlight_fields[2]=comments
    [16] => public-api.wordpress.com
    [17] => query=hello
    [18] => size=10
    [19] => sort=score_default
    [20] => w^BXg^rmuzai3iBGaV$*wANJ1n*Y3uzq:1:0
    [21] => 443
)
```

It looks like the Unix epoch is being treated as a string in the former and as a number in the latter.

I'm still investigating how these changes might affect the backward compatibility of older versions of the Jetpack Signature class.

Reply:
Ah ha! Thanks for clarifying. My suggestion to add `sort()` near the bottom of `->sign_request()` was a terrible one :) I misread the function.

We only want to sort the normalized query parameters, not the rest of the signature pieces. But we want the normalized query parameters to be flattened before we sort them. So we should probably move the flattening from `->sign_request()` to `->normalized_query_parameters()` and sort there.

Reply:
> Otherwise, we're sorting string and array values together, which may work but I don't trust to keep working.

Here's an example output of `$normalized_request_pieces` for a request containing an arrayed query string before this PR:
```php
[01-Sep-2020 21:01:46 UTC] [0c46-23884] Array
(
    [0] => w^BXg^rmuzai3iBGaV$*wANJ1n*Y3uzq:1:0
    [1] => 1598994106
    [2] => oO8QvSkiG5
    [3] =>
    [4] => GET
    [5] => public-api.wordpress.com
    [6] => 443
    [7] => /rest/v1.3/sites/160291064/search
    [8] => query=hello
    [9] => size=10
    [10] => sort=score_default
    [11] => highlight_fields[0]=title
    [12] => highlight_fields[1]=content
    [13] => highlight_fields[2]=comments
    [14] => fields[0]=date
    [15] => fields[1]=permalink.url.raw
    [16] => fields[2]=tag.name.default
    [17] => fields[3]=category.name.default
    [18] => fields[4]=post_type
    [19] => fields[5]=has.image
    [20] => fields[6]=shortcode_types
    [21] => fields[7]=image.url.raw
)
```

Here's the same output if we flatten and sort the query string pairs within `normalized_query_parameters`:
```php
(
    [0] => w^BXg^rmuzai3iBGaV$*wANJ1n*Y3uzq:1:0
    [1] => 1598994106
    [2] => oO8QvSkiG5
    [3] =>
    [4] => GET
    [5] => public-api.wordpress.com
    [6] => 443
    [7] => /rest/v1.3/sites/160291064/search
    [8] => fields[0]=date
    [9] => fields[1]=permalink.url.raw
    [10] => fields[2]=tag.name.default
    [11] => fields[3]=category.name.default
    [12] => fields[4]=post_type
    [13] => fields[5]=has.image
    [14] => fields[6]=shortcode_types
    [15] => fields[7]=image.url.raw
    [16] => highlight_fields[0]=title
    [17] => highlight_fields[1]=content
    [18] => highlight_fields[2]=comments
    [19] => query=hello
    [20] => size=10
    [21] => sort=score_default
)
```

As you can see, the existing implementation already relies on sorting string and array values together (i.e. arrayed query string values must be placed after string values upon being `[sort](https://github.com/Automattic/jetpack/blob/a426e63faafac04318689e743368e23914081b68/packages/connection/legacy/class-jetpack-signature.php#L307)`ed). Flattening and sorting the parameters in `normalized_query_parameters`, while straightforward, is guaranteed to break backward compatibility. 

I think it'd be best to preserve the `sort` changes originally proposed by this PR.

Aside: I'll start porting these changes to a separate PR.

Reply:
Thanks for the explanation. I agree we need to stick with the current method for now. (I wish we'd done things differently years ago :))

We may want to consider working on a path toward changing to a simpler sort though. PHP 8 changes the results of some comparisons (and so changes some sort orders). I am reasonably certain this code is not affected by those changes, but future extensions to the signature (for example, support for keys without values (e.g., no equals sign in one of those query parameter rows above)) definitely would be.